tools/debian/first_package_finder.py

```diff
@@ -0,0 +1,134 @@
+"""Functions to help find the first package version for a specific release"""
+import json
+from urllib import request
+from datetime import datetime
+import gzip
+
+import pandas as pd
+
+DEBIAN_RELEASE_VERSIONS_URL = 'https://debian.pages.debian.net/distro-info-data/debian.csv'
+
+
+def create_url(date: datetime, version: str = None):
+  """Create an url for snapshot.debian.org"""
+  base_url = 'https://snapshot.debian.org/archive/debian/{date}/dists/'
+  # `.gz` format always exist for all snapshots
+  source_path = '{version}/main/source/Sources.gz'
+
+  formatted_date = convert_datetime_to_str_datetime(date)
+
+  if version is None:
+    # return url for retrieving all versions for date
+    return base_url.format(date=formatted_date)
+  else:
+    return base_url.format(date=formatted_date) \
+           + source_path.format(version=version)
+
+
+def convert_datetime_to_str_datetime(input_datetime: datetime) -> str:
+  """Convert datetime object to debian snapshot url string"""
+  return input_datetime.isoformat().replace('-', '').replace(':', '') + 'Z'
+
+
+def create_codename_to_version() -> pd.DataFrame:
+  """Returns the codename to version mapping"""
+  with request.urlopen(DEBIAN_RELEASE_VERSIONS_URL) as csv:
+    df = pd.read_csv(csv, dtype=str)
+    # `series` appears to be `codename` but with no caps
+    df['sources'] = ''
+    df['first seen'] = ''
+    codename_to_version = df.set_index('series')
+    codename_to_version.loc['sid']['version'] = 'unstable'
+
+  return codename_to_version
+
+
+def parse_first_seen_dates(date: str) -> datetime:
+  """Parse first seen date in debian table to datetime"""
+  return datetime.strptime(date, '%Y-%m-%d %H:%M:%S')
+
+
+def fillout_first_seen(date: datetime, first_seen_dict: dict[str, datetime]):
+  """Fill out first seen version dict"""
+  with request.urlopen(create_url(date)) as result:
+    # Pandas will try to convert every table on the webpage to a dataframe.
```

Comment:
This is neat but potentially brittle..

Could we just use the first created date as listed on https://debian.pages.debian.net/distro-info-data/debian.csv?

Reply:
We can, the main issue is that we won't be getting the very first Source file, but probably the snapshot 24 - 36 hours later as I need to get the snapshot of the next day after the date listed there to reliably find it (though we can increase complexity and try a couple of different times around the listed date to make sure we get the first snapshot if this is very important).

Either case, I agree the current method is a bit brittle, so I'll put the current method in the alternative's we've considered part of the design doc.

Reply:
Just implemented it using debian.csv. Because debian snapshot api will always retrieve the previous snapshot for each date that is passed in, if there is a break in snapshots around when a release is made (for example when squeeze is released, there was over a week without snapshots), it won't find the first snapshot. 

This could potentially be fixed by looking forward a couple days until it does find a snapshot containing the release.

Reply:
OK, let's go with this. retrying by walking a few days doesn't sound too bad, and is much more reliable than parsing the html table.